camus-kafka-coders/src/main/java/com/linkedin/camus/etl/kafka/coders/AvroMessageDecoder.java

```diff
@@ -0,0 +1,182 @@
+  private Schema latestSchema;
+  private int latestVersion;
+  private String topic;
+  private boolean isNew;
+
+  public AvroMessageDecoder(SchemaRegistryClient schemaRegistry) {
+    this.schemaRegistry = schemaRegistry;
+  }
+
+  public AvroMessageDecoder() {
+
+  }
+
+  @Override
+  public void init(Properties props, String topicName) {
+    super.init(props, topicName);
+
+    decoderFactory = DecoderFactory.get();
+    if (props == null) {
+      throw new IllegalArgumentException("Missing schema registry url!");
+    }
+    String baseUrl = props.getProperty(SCHEMA_REGISTRY_URL);
+    if (baseUrl == null) {
+      throw new IllegalArgumentException("Missing schema registry url!");
+    }
+    String maxSchemaObject = props.getProperty(
+        MAX_SCHEMAS_PER_SUBJECT, DEFAULT_MAX_SCHEMAS_PER_SUBJECT);
+    if (schemaRegistry == null) {
+      schemaRegistry = new CachedSchemaRegistryClient(baseUrl, Integer.parseInt(maxSchemaObject));
+    }
+    this.isNew = Boolean.parseBoolean(props.getProperty(IS_NEW_PRODUCER, "true"));
+    this.topic = topicName;
+  }
+
+  private ByteBuffer getByteBuffer(byte[] payload) {
+    ByteBuffer buffer = ByteBuffer.wrap(payload);
+    byte magic = buffer.get();
+    logger.debug("MAGIC BYTE" + magic);
+    if (magic != MAGIC_BYTE) {
+      throw new MessageDecoderException("Unknown magic byte!");
+    }
+    return buffer;
+  }
+
+  private String constructSubject(String topic, Schema schema, boolean isNewProducer) {
+    if (isNewProducer) {
+      return topic + "-value";
+    } else {
+      return schema.getName() + "-value";
+    }
+  }
+
+  private Object deserialize(byte[] payload) throws MessageDecoderException {
+    try {
+      ByteBuffer buffer = getByteBuffer(payload);
+      int id = buffer.getInt();
+      Schema schema = schemaRegistry.getByID(id);
+      if (schema == null)
+        throw new IllegalStateException("Unknown schema id: " + id);
+      logger.debug(schema.toString());
+      String subject = constructSubject(topic, schema, isNew);
+      logger.debug("Subject = " + subject);
+      if (latestSchema == null) {
+        latestSchema = parser.parse(schemaRegistry.getLatestSchema(subject));
+        latestVersion = schemaRegistry.getLatestVersion(subject);
+      }
+
+      int version = schemaRegistry.getVersion(subject, schema);
+      if (version > latestVersion) {
+        String errorMsg = String.format(
+            "Producer schema is newer than the schema known to Camus");
+        throw new MessageDecoderException(errorMsg);
+      }
+
+      int length = buffer.limit() - 1 - idSize;
+      if (schema.getType().equals(Schema.Type.BYTES)) {
+        byte[] bytes = new byte[length];
+        buffer.get(bytes, 0, length);
+        return bytes;
+      }
+      int start = buffer.position() + buffer.arrayOffset();
+      DatumReader<Object> reader = new GenericDatumReader<Object>(latestSchema);
+      Object object =
```

Comment:
I think you will need to use the following api since the schema of the message may not be the latest and we want to project to the latest schema.

genericdatumreader(schema writer, schema reader)